lib/common/xml.c

```diff
@@ -1138,6 +1073,60 @@ pcmk__xml_parse_file(const char *filename)
     return xml;
 }
 
+/*!
+ * \internal
+ * \brief Parse XML from a string
+ *
+ * \param[in] input  String to parse
+ *
+ * \return XML tree parsed from the given string; may be \c NULL or only partial
+ *         on error
+ */
+xmlNode *
+pcmk__xml_parse_string(const char *input)
+{
+    xmlNode *xml = NULL;
+    xmlDoc *output = NULL;
+    xmlParserCtxt *ctxt = NULL;
+    const xmlError *last_error = NULL;
+
+    if (input == NULL) {
+        crm_err("Can't parse NULL input");
+        return NULL;
+    }
+
+    // Create a parser context
+    ctxt = xmlNewParserCtxt();
+    CRM_CHECK(ctxt != NULL, return NULL);
+
+    xmlCtxtResetLastError(ctxt);
+    xmlSetGenericErrorFunc(ctxt, pcmk__log_xmllib_err);
+
+    parse_xml_recover(&output, xmlCtxtReadDoc, ctxt, (pcmkXmlStr) input, NULL,
+                      NULL);
+
+    if (output != NULL) {
+        xml = xmlDocGetRootElement(output);
+    }
+
+    last_error = xmlCtxtGetLastError(ctxt);
+    if (last_error != NULL) {
+        crm_err("Couldn't %sparse XML from string: %s "
+                CRM_XS " (domain=%d, level=%d, code=%d)",
+                ((xml != NULL)? "fully " : ""), last_error->message,
+                last_error->domain, last_error->level, last_error->code);
+
+        crm_info("XML parse error: %s", input);
```

Comment:
I'm thinking we may want this at error level, but I don't feel strongly about it. I had set it to info because input might be a large string.